tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,753 @@
+	// The int key in the map is the file number on the disk.
+	mmappedChunkFiles map[int]ByteSlice
+	closers           map[int]io.Closer // Closers for resources behind the byte slices.
+	mmappedChunksMtx  sync.RWMutex
+	pool              chunkenc.Pool
+	size              int64 // The total size of bytes in the reader.
+
+	closed bool
+}
+
+const (
+	// MintMaxtSize is the size of the mint/maxt for head chunk file and chunks.
+	MintMaxtSize = 8
+	// SeriesRefSize is the size of series reference on disk.
+	SeriesRefSize = 8
+	// HeadChunkFileHeaderSize is the total size of the header for the head chunk file.
+	HeadChunkFileHeaderSize = SegmentHeaderSize + 2*MintMaxtSize
+	// HeaderMintOffset is the offset where the first byte of MinT for head chunk file exists.
+	HeaderMintOffset = SegmentHeaderSize
+	// HeaderMaxtOffset is the offset where the first byte of MaxT for head chunk file exists.
+	HeaderMaxtOffset = HeaderMintOffset + MintMaxtSize
+	// MaxHeadChunkFileSize is the max size of a head chunk file.
+	// Setting size to the max int32 as setting it to max int64 crashes 64 systems too.
+	MaxHeadChunkFileSize = math.MaxInt32
+	// CRCSize is the size of crc32 sum on disk.
+	CRCSize = 4
+	// MaxHeadChunkMetaSize is the max size of an mmapped chunks minus the chunks data.
+	// Max because the uvarint size can be smaller.
+	MaxHeadChunkMetaSize = SeriesRefSize + 2*MintMaxtSize + ChunksFormatVersionSize + MaxChunkLengthFieldSize + CRCSize
+)
+
+// NewChunkDiskMapper returns a new writer against the given directory
+// using the default head chunk file duration.
+func NewChunkDiskMapper(dir string, pool chunkenc.Pool) (*ChunkDiskMapper, error) {
+	return newChunkDiskMapper(dir, DefaultHeadChunkFileMaxTimeRange, pool)
+}
+
+func newChunkDiskMapper(dir string, maxFileDuration int64, pool chunkenc.Pool) (*ChunkDiskMapper, error) {
+	if maxFileDuration <= 0 {
+		maxFileDuration = DefaultHeadChunkFileMaxTimeRange
+	}
+
+	if err := os.MkdirAll(dir, 0777); err != nil {
+		return nil, err
+	}
+	dirFile, err := fileutil.OpenDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &ChunkDiskMapper{
+		dir:       dirFile,
+		maxFileMs: maxFileDuration,
+		pool:      pool,
+		crc32:     newCRC32(),
+	}
+
+	for i := 0; i < inBufferShards; i++ {
+		m.inBufferChunks[i] = make(map[uint64]chunkenc.Chunk)
+	}
+
+	return m, m.openMMapFiles()
+}
+
+func (cdm *ChunkDiskMapper) openMMapFiles() (returnErr error) {
+	bs := map[int]ByteSlice{}
+	cs := map[int]io.Closer{}
+	defer func() {
+		if returnErr != nil {
+			var merr tsdb_errors.MultiError
+			merr.Add(returnErr)
+			merr.Add(closeAllFromMap(cs))
+			returnErr = merr.Err()
+		}
+	}()
+
+	files, err := sequenceFilesMap(cdm.dir.Name())
+	if err != nil {
+		return err
+	}
+	if cdm.pool == nil {
+		cdm.pool = chunkenc.NewPool()
+	}
+
+	seqs := make([]int, 0, len(files))
+	for seq, fn := range files {
+		f, err := fileutil.OpenMmapFile(fn)
+		if err != nil {
+			return errors.Wrap(err, "mmap files")
+		}
+		cs[seq] = f
```

Comment:
Slices might be better because:
* You don't need to rewrite + sort those maps all the time
* Sometimes because of the above, you just not sort, and do more inefficient stuff (e.g search)
* With slice you can then reuse underlying array when you reuse (e.g in bufferedChunks)

Reply:
> We can operate on just slice indexes, right? Will be much cleaner and deterministic, no?

The sequence does not always start at 0. After a point once we start deleting old files, it will never start from 0, and keeping track of offset and using a slice would be more buggy and complex. Map feels simpler here. 

Plus, there won't be a lot of files, so I would not worry about re-use and search here, maps would be quick enough. And we only sort the files in `IterateAllChunks` which will be done only once on start, so I would not worry about that too.